test/sanity/code-smell/rstcheck.py

```diff
@@ -4,59 +4,169 @@
 
-    pattern = re.compile(r'^(?P<path>[^:]*):(?P<line>[0-9]+): \((?P<level>INFO|WARNING|ERROR|SEVERE)/[0-4]\) (?P<message>.*)$')
+def _map_doc_paths_to_extra_lint_args(rst_doc_paths: t.Tuple[str]):
+    generic_doc_indexes: t.List[int] = []
+    sphinx_cli_doc_indexes: t.List[int] = []
 
-    results = parse_to_list_of_dict(pattern, process.stderr.decode(encoding))
+    sphinx_cli_doc_path_prefix = 'docs/docsite/rst/cli/ansible'
+
+    ignored_rst_substitutions = (
+        'br',
+    )
+    ignored_in_tree_sphinx_directives = (
+        'ansible-cmd-docs',
+    )
+
+    generic_doc_extra_rstlint_args: t.List[str] = [
+        '--ignore-substitutions', ','.join(ignored_rst_substitutions),
+    ]
+    sphinx_cli_doc_extra_rstlint_args: t.List[str] = [
+        *generic_doc_extra_rstlint_args,
+        '--ignore-directives', ','.join(ignored_in_tree_sphinx_directives),
+    ]
+
+    for doc_index, doc_path in enumerate(rst_doc_paths):
+        is_sphinx_cli_doc = doc_path.startswith(sphinx_cli_doc_path_prefix)
+        target_indexes_list = (
+            sphinx_cli_doc_indexes if is_sphinx_cli_doc
+            else generic_doc_indexes
+        )
+        target_indexes_list.append(doc_index)
+
+    generic_doc_paths_generator = (
+        rst_doc_paths[doc_index] for doc_index in generic_doc_indexes
+    )
+    sphinx_cli_paths_generator = (
+        rst_doc_paths[doc_index] for doc_index in sphinx_cli_doc_indexes
+    )
+    return zip(
+        (generic_doc_extra_rstlint_args, sphinx_cli_doc_extra_rstlint_args),
+        (generic_doc_paths_generator, sphinx_cli_paths_generator),
+    )
 
-    for result in results:
-        print('%s:%s:%s: %s' % (result['path'], result['line'], 0, result['message']))
 
+def _lint_rst_docs(
+        rst_doc_paths: t.Iterable[str],
+        rstlint_extra_cli_args: t.List[str],
+):
+    unwinded_rst_doc_paths = tuple(rst_doc_paths)
+    if not unwinded_rst_doc_paths:  # calling `rstcheck` not passing docs fails
+        return
 
-def parse_to_list_of_dict(pattern, value):
+    rstlint_cmd = (
+        sys.executable,
+        '-c', 'import rstcheck; rstcheck.main();',
+        '--report', 'warning',
+        *rstlint_extra_cli_args,
+        *unwinded_rst_doc_paths,
+    )
+    rstlint_cmd_text = ' '.join(rstlint_cmd)
+
+    try:
+        rstlint_result = subprocess.run(
+            rstlint_cmd,
+            check=True,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,  # Has a `text` alias since Python 3.7
+        )
+    except subprocess.CalledProcessError as proc_err:
+        raise RuntimeError(
+            'Oops... Got unexpected error output from '
+            f'``{rstlint_cmd_text !s}``:\n\n{proc_err.stderr !s}',
+        ) from proc_err
+
+    if rstlint_result.stdout:
+        raise RuntimeError(
+            f'Oops... Got unexpected output from ``{rstlint_cmd_text !s}``:'
```

Comment:
Just pointing this out, take it or leave it but I'm not sure if "Oops..." is meaningful. It's fun and friendly but does it risk confusion, especially for non-native English speakers? IMO it's better to drop the oops part.

Reply:
I think I'd like to keep this. Unless @mattclay would also want it out. This case is something that should never happen IRL...